### PR TITLE
Fast sam transcoder

### PR DESCRIPTION
**Pull request by MueFab**
# Fast SAM Transcoding #
## Optimizations ##
This pull request implements a few optimizations to speed up the sam-to-mgrec transcoding:

* Reducing disk I/O by sorting the reads before writing them to disk instead of sorting in a separate step
* Bypassing the BitWriter bit logic when writing mgrec records (all fields are byte aligned)
* Using a heap when merging sorted files instead of naive comparison
* Introducing multithreaded transcoding
* Fixing the broken move constructor of the Record class
* Preallocate fields with the correct sizes if known in advance

All in all, around **5.2x speedup** can be observed with all optimizations applied. A 97GB SAM file can be transcoded in 21 minutes instead of 111 minutes.

## More Features ##
* BAM files as input files (BAM output not yet supported)
* Input and output from / to standard input and output. This is enabled by using a dash as file name (e.g. "-.mgrec") and can directly pipe the output to an other application, avoiding disk I/O (e.g. connecting the transcoder directly to the genie compression)

### CODE REVIEW

File: src/genie/util/bitreader.impl.h [hunk 1]
```diff
@@ -28,11 +40,40 @@ T BitReader::read(uint8_t s) {
...
 
+template <typename T, size_t SIZE, typename>
+T BitReader::readBypassBE() {
+    static_assert(SIZE > 0, "SIZE == 0");
+    static_assert(SIZE <= sizeof(T), "SIZE > sizeof(T)");
+    T ret = static_cast<T>(0);
+    istream.read(reinterpret_cast<char*>(&ret), SIZE);
+
+    // Swap Endianness if necessary
+    if (SIZE > 1) {
```
Comment (sXperfect) [inline src/genie/util/bitreader.impl.h:51]: Why do we need to swap the Endian for readBypassBE?

File: src/genie/core/record/alignment-box.cc [hunk 2]
```diff
@@ -30,15 +30,16 @@ void AlignmentBox::write(util::BitWriter& writer) const {
...
+    : splitAlignmentInfo(type == ClassType::CLASS_HM ? 0 : number_of_template_segments - 1) {
+    mapping_pos = reader.readBypassBE<uint64_t, 5>();
     alignment = Alignment(as_depth, reader);
 
     if (type == ClassType::CLASS_HM) {
         return;
     }
     for (size_t tSeg = 1; tSeg < number_of_template_segments; tSeg++) {
-        splitAlignmentInfo.push_back(AlignmentSplit::factory(as_depth, reader));
+        splitAlignmentInfo[tSeg - 1] = AlignmentSplit::factory(as_depth, reader);
```
Comment (sXperfect) [inline src/genie/core/record/alignment-box.cc:42]: Why FIFO with offset 1?

File: src/apps/genie/transcode-sam/utils.cc [hunk 3]
```diff
@@ -21,6 +21,9 @@ bool compare(const genie::core::record::Record& r1, const genie::core::record::R
     if (r2.getAlignments().empty()) {
         return true;
     }
+    if (r1.getAlignmentSharedData().getSeqID() != r2.getAlignmentSharedData().getSeqID()) {
```
Comment (sXperfect) [inline src/apps/genie/transcode-sam/utils.cc:24]: Why do we have this additional condition for sorting?

File: src/genie/core/record/record.h [hunk 4]
```diff
@@ -37,14 +37,14 @@ namespace record {
 class Record {
  private:
     uint8_t number_of_template_segments{};    //!< @brief
+    std::vector<Segment> reads;               //!< @brief
```
Comment (sXperfect) [inline src/genie/core/record/record.h:40]: You just reorder things here right?

File: src/genie/core/record/alignment-box.cc (hunk 2 above)
Comment (sXperfect) [reply src/genie/core/record/alignment-box.cc:42]: *Why add to front and not to the back?

Comment (sXperfect) [reply src/genie/core/record/alignment-box.cc:42]: *Sorry, i see what you did there.
You reserve the memory first then add the elements

File: src/genie/util/bitreader.impl.h (hunk 1 above)
Comment (MueFab) [reply src/genie/util/bitreader.impl.h:51]: x86 architecture is little endian, but data fields in MPEG-G are written out big endian. This is an implication required by part 2, 6.2:
"The bit order of syntax fields in the syntax tables is specified to start with the
MSB  and  proceed to the  LSB"-
The regular bitwriter does this automatically, when using bypass we need to change the order manually instead,

File: src/genie/core/record/alignment-box.cc (hunk 2 above)
Comment (MueFab) [reply src/genie/core/record/alignment-box.cc:42]: Only the second alignment is a split alignment, the first alignment is already included in the "Alignment Box". However number_of_template_segments includes both of them. See the structure of MPEG-G record in part 2

File: src/apps/genie/transcode-sam/utils.cc (hunk 3 above)
Comment (MueFab) [reply src/apps/genie/transcode-sam/utils.cc:24]: One of the optimizations in this pull request is to sort by sequence and position at the same time. We have no separate files for chromosomes anymore - instead the chromosomes are just considered to be part of the position.

1. Chromosomes differ, use chromosomes to infer the order of records
2. If chromosomes are different, use position on the chromosome instead

File: src/genie/core/record/record.h (hunk 4 above)
Comment (MueFab) [reply src/genie/core/record/record.h:40]: Yes it now in the order of the binary representation of the record

**Review by sXperfect — APPROVED** [review]

**Pull request merged**